Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp

```diff
@@ -277,78 +269,116 @@ void MediaRecorderPrivateEncoder::appendAudioSampleBuffer(const PlatformAudioDat
+
     {
-        if (!m_ringBufferLock.tryLock())
-            return;
-        Locker locker { AdoptLock, m_ringBufferLock };
-        if (!m_ringBuffer)
+        Locker locker { m_ringBufferLock };
+        if (m_ringBuffers.isEmpty() || !m_ringBuffers.last())
             return;
-        m_ringBuffer->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_lastEnqueuedAudioSampleCount);
+        m_ringBuffers.last()->store(downcast<WebAudioBufferList>(data).list(), sampleCount, m_currentAudioSampleCount);
     }
-    queueSingleton().dispatch([weakThis = ThreadSafeWeakPtr { *this }, sampleCount, lastEnqueued = m_lastEnqueuedAudioSampleCount.load()] {
+
+    queueSingleton().dispatch([weakThis = ThreadSafeWeakPtr { *this }, currentAudioTime, sampleCount, currentAudioSampleCount = m_currentAudioSampleCount] {
         if (RefPtr protectedThis = weakThis.get())
-            protectedThis->audioSamplesAvailable(sampleCount, lastEnqueued);
+            protectedThis->audioSamplesAvailable(currentAudioTime, sampleCount, currentAudioSampleCount);
     });
+
+    m_currentAudioSampleCount += sampleCount;
 }
 
-void MediaRecorderPrivateEncoder::audioSamplesAvailableStarted(const AudioStreamBasicDescription& description)
+void MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged(const AudioStreamBasicDescription& description)
 {
     assertIsCurrent(queueSingleton());
 
-    ASSERT(!m_audioFormatDescription);
+    if (!m_originalOutputDescription) {
+        AudioStreamBasicDescription outputDescription = { };
+        outputDescription.mFormatID = m_audioCodec;
+        outputDescription.mChannelsPerFrame = description.mChannelsPerFrame;
+        outputDescription.mSampleRate = description.mSampleRate;
+        m_originalOutputDescription = outputDescription;
+    }
+
     CMFormatDescriptionRef newFormat = nullptr;
     if (auto error = PAL::CMAudioFormatDescriptionCreate(kCFAllocatorDefault, &description, 0, nullptr, 0, nullptr, nullptr, &newFormat)) {
         RELEASE_LOG_ERROR(MediaStream, "MediaRecorderPrivateEncoder::audioSamplesAvailable: CMAudioFormatDescriptionCreate failed with %u", error);
+        m_hadError = true;
         return;
     }
     m_audioFormatDescription = adoptCF(newFormat);
+
+    if (m_audioCompressor) {
+        audioCompressor()->finish();
+        m_formatChangedOccurred = true;
+    }
+
+    m_audioCompressor = AudioSampleBufferCompressor::create(compressedAudioOutputBufferCallback, this, m_audioCodec, *m_originalOutputDescription);
+    if (!m_audioCompressor) {
+        RELEASE_LOG_ERROR(MediaStream, "MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged: creation of compressor failed");
+        m_hadError = true;
+        return;
+    }
+
+    if (m_audioBitsPerSecond)
+        audioCompressor()->setBitsPerSecond(m_audioBitsPerSecond);
+
+    Locker locker { m_ringBufferLock };
+    if (m_ringBuffers.size() > 1)
+        m_ringBuffers.removeFirst();
```

Comment:
I am not sure to really understand why we use a Deque here. Cannot we just use two std::unique_ptr ring buffers, X and Y?

X would only be used to read and is only set in the queue.
Y would be created where audio samples come from.

On the read side, when we detect that Y is not null, we move Y into X.
On the write side, when Y is not null, we write into Y, otherwise, we write into X.

Reply:
It seems that you missed how the Deque is used. 
Say we have 3 format changes occurring.
A | B | C
When A is received, RingBuffer A is created and an `audioSamplesDescriptionChanged` is dispatched to the queue.
Samples with format A, they are added to RingBuffer A, and are all dispatched to `audioSamplesAvailable` to the queue.

Do the same for B and C; none of the tasks dispatched has been run yet. All content has been added to their respective ring buffer.

`audioSamplesDescriptionChanged` for A now runs, it takes the pointer to the first RingBuffer and place it in `m_currentRingBuffer` ; having `m_currentRingBuffer` allows for the reader to pull from the RingBuffer without using the need for the lock.

Now `audioSamplesAvailable` for A is run, fetch the samples from RingBuffer A.
---
`audioSamplesDescriptionChanged` for B now runs, we can pop the first ring buffer from the deque (which is RingBuffer A), there are no A frames in flight/pending, the next ones will be of B format.

Rinse & repeat.

While it's an unlikely scenario, it is possible. The use of an unlimited sized queue is required.